### PR TITLE
fix: Fixing pinot query generation for date format conversion from python datetime format to java simple date format

### PR DESCRIPTION
**Pull request by xiangfu0**
### SUMMARY
Fixing Pinot query generation converting DateTime format from python DateTime format to Java SimpleDate Format.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
After：
Generated query and chart:
```
SELECT DATETIMECONVERT(FlightDate, '1:SECONDS:SIMPLE_DATE_FORMAT:yyyy-MM-dd', '1:SECONDS:SIMPLE_DATE_FORMAT:yyyy-MM-dd', '1:DAYS'),
       COUNT(*)
FROM "airlineStats"."airlineStats"
GROUP BY DATETIMECONVERT(FlightDate, '1:SECONDS:SIMPLE_DATE_FORMAT:yyyy-MM-dd', '1:SECONDS:SIMPLE_DATE_FORMAT:yyyy-MM-dd', '1:DAYS')
LIMIT 10000;
```
image

### TEST PLAN
Adding unit test for query generation.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/db_engine_specs/pinot.py [hunk 1]
```diff
@@ -86,7 +86,9 @@ def get_timestamp_expr(
                 python_pattern,
                 java_pattern,
             ) in cls._python_to_java_time_patterns.items():
-                java_date_format.replace(python_pattern, java_pattern)
+                java_date_format = java_date_format.replace(
+                    python_pattern, java_pattern
+                )
```
Comment (villebro) [inline superset/db_engine_specs/pinot.py:91]: Line 84 won't work correctly for `pdf = None` (I assume `pdf` will always be defined, but it doesn't look right). Let's fix that now, too, something like
```python
            if pdf:
                java_date_format = pdf
                for (
                    python_pattern,
                    java_pattern,
                ) in cls._python_to_java_time_patterns.items():
                    java_date_format = java_date_format.replace(
                        python_pattern, java_pattern
                    )
```

**Review by villebro — COMMENTED** [review]
Thanks for fixing again! Quick bycatch comment

**Comment by villebro** [comment]
@fx19880617 btw you may want to cherry pick this into your installation, as I believe this fix also affects Pinot: https://github.com/apache/superset/pull/13138

File: superset/db_engine_specs/pinot.py (hunk 1 above)
Comment (fx19880617) [reply superset/db_engine_specs/pinot.py]: Thanks! Fixed.

**Comment by fx19880617** [comment]
> @fx19880617 btw you may want to cherry pick this into your installation, as I believe this fix also affects Pinot: #13138

Sure, is it on the latest docker image?

**Review by villebro — COMMENTED** [review]
For completeness, could we also add an assertion for missing `pdf` and invalid time grain spec?

**Comment by fx19880617** [comment]
> For completeness, could we also add an assertion for missing `pdf` and invalid time grain spec?

it's already there on line 72 for pdf check and line 99 for time_grain check.

**Comment by villebro** [comment]
> it's already there on line 72 for pdf check and line 99 for time_grain check.

Yes, I meant a simple unit test to test for those lines. The reason I'm asking is Pinot is a fairly different connector compared to others, so I'd like to make sure we have good coverage for it to guard against regressions.

File: tests/db_engine_specs/pinot_tests.py [hunk 2]
```diff
@@ -32,10 +32,48 @@ def test_pinot_time_expression_sec_one_1d_grain(self):
...
         expr = PinotEngineSpec.get_timestamp_expr(col, "epoch_s", "P1M")
         result = str(expr.compile())
         self.assertEqual(
             result, "DATETRUNC('month', tstamp, 'SECONDS')",
         )  # noqa
+
+    def test_invalid_get_time_expression_arguments(self):
+        with self.assertRaises(NotImplementedError) as context:
+            PinotEngineSpec.get_timestamp_expr(column("tstamp"), None, "P1M")
+        self.assertEqual("Empty date format for 'tstamp'", str(context.exception))
```
Comment (villebro) [inline tests/db_engine_specs/pinot_tests.py:71]: Is this indented correctly? I'm fine removing the string comparison, testing for the raise is adequate IMO.

File: tests/db_engine_specs/pinot_tests.py [hunk 3]
```diff
@@ -32,10 +32,48 @@ def test_pinot_time_expression_sec_one_1d_grain(self):
...
+
+        with self.assertRaises(NotImplementedError) as context:
+            PinotEngineSpec.get_timestamp_expr(
+                column("tstamp"), "epoch_s", "invalid_grain"
+            )
+        self.assertEqual(
+            "No pinot grain spec for 'invalid_grain'", str(context.exception)
+        )
```
Comment (villebro) [inline tests/db_engine_specs/pinot_tests.py:79]: Same here

File: tests/db_engine_specs/pinot_tests.py (hunk 2 above)
Comment (fx19880617) [reply tests/db_engine_specs/pinot_tests.py:71]: I tried to use VS Code pylint plugin and black to format the python file, but both don't pass the python lint check here :(
The black cmd I'm running is 
```black -t py37 tests/db_engine_specs/pinot_tests.py```

Comment (fx19880617) [reply tests/db_engine_specs/pinot_tests.py:71]: I also tried:
``` self.assertRaises(
            NotImplementedError,
           PinotEngineSpec.get_timestamp_expr(column("tstamp"), None, "P1M"),
        )
```
but the test doesn't pass for this, hence I capture the exception and compare the string:
```
================================================================================================= FAILURES ==================================================================================================
_____________________________________________________________________ TestPinotDbEngineSpec.test_invalid_get_time_expression_arguments ______________________________________________________________________

self = <tests.db_engine_specs.pinot_tests.TestPinotDbEngineSpec testMethod=test_invalid_get_time_expression_arguments>

    def test_invalid_get_time_expression_arguments(self):
        self.assertRaises(
            NotImplementedError,
>           PinotEngineSpec.get_timestamp_expr(column("tstamp"), None, "P1M"),
        )

tests/db_engine_specs/pinot_tests.py:71:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'superset.db_engine_specs.pinot.PinotEngineSpec'>, col = <sqlalchemy.sql.elements.ColumnClause at 0x11f54a5e0; tstamp>, pdf = None, time_grain = 'P1M', type_ = None

    @classmethod
    def get_timestamp_expr(
        cls,
        col: ColumnClause,
        pdf: Optional[str],
        time_grain: Optional[str],
        type_: Optional[str] = None,
    ) -> TimestampExpression:
        if not pdf:
>           raise NotImplementedError(f"Empty date format for '{col}'")
E           NotImplementedError: Empty date format for 'tstamp'

superset/db_engine_specs/pinot.py:72: NotImplementedError
```

Comment (villebro) [reply tests/db_engine_specs/pinot_tests.py]: I tweaked the tests slightly and pushed to the branch, I hope you don't mind (if you don't agree with the changes please feel free to `git reset --hard HEAD~1`!)

**Review by villebro — APPROVED** [review]

**Comment by villebro** [comment]
@fx19880617 let me know if my last change is ok by you - if so I can merge

**Comment by fx19880617** [comment]
> @fx19880617 let me know if my last change is ok by you - if so I can merge

LGTM, thanks for fixing it @villebro

**Pull request merged**